src/ucp/tag/offload.c

```diff
@@ -429,61 +391,35 @@ static ucs_status_t ucp_tag_offload_eager_zcopy(uct_pending_req_t *self)
                                     ucp_proto_am_zcopy_req_complete);
 }
 
-static size_t ucp_tag_offload_rndv_pack_rkey(ucp_request_t *sreq, ucp_lane_index_t lane,
-                                             void *rkey_buf, uint16_t *flags)
-{
-    ucp_ep_h ep = sreq->send.ep;
-    ucs_status_t status;
-
-    ucs_assert(UCP_DT_IS_CONTIG(sreq->send.datatype));
-
-    /* Check if the sender needs to register the send buffer -
-     * is its datatype contiguous and does the receive side need it */
-    if (ucp_ep_md_attr(ep, lane)->cap.flags & UCT_MD_FLAG_NEED_RKEY) {
-        status = ucp_request_send_buffer_reg(sreq, lane);
-        ucs_assert_always(status == UCS_OK);
-
-        /* if the send buffer was registered, send the rkey */
-        UCS_PROFILE_CALL(uct_md_mkey_pack, ucp_ep_md(ep, lane),
-                         sreq->send.state.dt.dt.contig[0].memh, rkey_buf);
-        *flags |= UCP_RNDV_RTS_FLAG_PACKED_RKEY;
-        return ucp_ep_md_attr(ep, lane)->rkey_packed_size;
-    }
-
-    return 0;
-}
-
 ucs_status_t ucp_tag_offload_sw_rndv(uct_pending_req_t *self)
 {
     ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
     ucp_ep_t *ep       = req->send.ep;
     ucp_sw_rndv_hdr_t *rndv_hdr;
-    ucp_sw_rndv_ext_hdr_t *ext_rndv_hdr;
+    ucp_rndv_rts_hdr_t *rndv_rts_hdr;
     unsigned rndv_hdr_len;
-    size_t packed_rkey;
 
     if (UCP_DT_IS_CONTIG(req->send.datatype) &&
         (req->send.length > ucp_ep_config(ep)->tag.offload.max_rndv_zcopy)) {
-        packed_rkey  = ucp_ep_md_attr(ep, req->send.lane)->rkey_packed_size;
-        rndv_hdr_len = sizeof(ucp_sw_rndv_ext_hdr_t) + packed_rkey;
-        ext_rndv_hdr = ucs_alloca(rndv_hdr_len);
-        ext_rndv_hdr->address     = (uintptr_t)req->send.buffer;
-        ext_rndv_hdr->super.flags = 0;
-        ucp_tag_offload_rndv_pack_rkey(req, req->send.lane, ext_rndv_hdr + 1,
-                                       &ext_rndv_hdr->super.flags);
-        rndv_hdr = &ext_rndv_hdr->super;
+        /* send RTS to allow fallback to SW RNDV on receiver */
+        rndv_hdr_len = ucp_rndv_rts_packed_len(ep);
+        rndv_rts_hdr = ucs_alloca(rndv_hdr_len);
+        ucs_assert(rndv_rts_hdr != NULL);
+        ucp_tag_rndv_rts_pack(rndv_rts_hdr, req);
+        return uct_ep_tag_rndv_request(ep->uct_eps[req->send.lane], req->send.tag,
+                                       rndv_rts_hdr, rndv_hdr_len);
     } else {
         rndv_hdr_len    = sizeof(ucp_sw_rndv_hdr_t);
         rndv_hdr        = ucs_alloca(rndv_hdr_len);
         rndv_hdr->flags = 0;
-    }
 
-    rndv_hdr->super.sender_uuid = ep->worker->uuid;
-    rndv_hdr->super.reqptr      = (uintptr_t)req;
-    rndv_hdr->length            = req->send.length;
+        rndv_hdr->super.sender_uuid = ep->worker->uuid;
```

Comment:
why don't you pack the whole RTS here as well? That will not work, because you treat it as RTS on the receiver

Reply:
fixed